### PR TITLE
Register username tx

### PR DESCRIPTION
**Pull request by webmaster128**
Expected to fail with

```
1. BnsConnection : can register a username
   Error: {
     "checkTx": {
       "log": "Id is invalid",
       "code": 501
     },
     "deliverTx": {
       "code": 0
     }
   }
```

### CODE REVIEW

File: packages/iov-bcp-types/types/transactions.d.ts [hunk 1]
```diff
@@ -61,4 +68,9 @@ export interface SwapTimeoutTx extends BaseTx {
     readonly kind: TransactionKind.SwapTimeout;
     readonly swapId: SwapIdBytes;
 }
-export declare type UnsignedTransaction = SendTx | SetNameTx | SwapOfferTx | SwapCounterTx | SwapClaimTx | SwapTimeoutTx;
+export interface RegisterUsernameTx extends BaseTx {
+    readonly kind: TransactionKind.RegisterUsername;
+    readonly username: string;
+    readonly addresses: Map<ChainId, Address>;
```
Comment (ethanfrey) [inline packages/iov-bcp-types/types/transactions.d.ts]: I kept reading that tagged types didn't work as keys in a map.
So I was always using sting.
Nice if you can keep that safety here

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 2]
```diff
@@ -321,6 +322,43 @@ describe("BnsConnection", () => {
...
+    const identity = await profile.createIdentity(wallet.id, HdPaths.simpleAddress(0));
+
+    // Create and send registration
+    const address = keyToAddress(identity.pubkey);
+    const username = `testuser`; // _${Math.random()}`;
+    const registration: RegisterUsernameTx = {
+      kind: TransactionKind.RegisterUsername,
+      chainId: chainId,
+      signer: identity.pubkey,
+      addresses: new Map([[chainId, address]]),
```
Comment (ethanfrey) [inline packages/iov-bns/src/bnsconnection.spec.ts]: This is a good question for me, what the address should be here, why bytes is human readable encoding, etc...

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 3]
```diff
@@ -321,6 +322,43 @@ describe("BnsConnection", () => {
...
+      username: username,
+    };
+    const nonce = await getNonce(connection, address);
+    const signed = await profile.signTransaction(wallet.id, identity, registration, bnsCodec, nonce);
+    const txBytes = bnsCodec.bytesToPost(signed);
+    await connection.postTx(txBytes);
+
+    // Find registration transaction
+    const searchResult = await connection.searchTx({ tags: [bnsFromOrToTag(address)] });
```
Comment (ethanfrey) [inline packages/iov-bns/src/bnsconnection.spec.ts]: Do not expect this to be found, this is only if the balance changes at the given address.

`bnsNonceTag` should find it, as it detects anything that was signed by that address

File: packages/iov-bns/src/bnsconnection.spec.ts (hunk 2 above)
Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.spec.ts]: Good question indeed. What happens when one chain supports different address types? I tend to think we should store the user's choice.

For some broken chains out there, there is no proper binary address.

**Comment by webmaster128** [comment]
okay, this is green locally. @ethanfrey may I get your approval?

**Pull request merged**